src/CommunityToolkit.Maui/Behaviors/MaskedBehavior.shared.cs

```diff
@@ -35,127 +36,135 @@ public class MaskedBehavior : BaseBehavior<InputView>
-
 	/// <inheritdoc />
-	protected override void OnViewPropertyChanged(object? sender, PropertyChangedEventArgs e)
+	protected override async void OnViewPropertyChanged(InputView sender, PropertyChangedEventArgs e)
 	{
 		base.OnViewPropertyChanged(sender, e);
 
 		if (e.PropertyName == InputView.TextProperty.PropertyName)
 		{
-			OnTextPropertyChanged();
+			await OnTextPropertyChanged();
 		}
 	}
 
-	void OnTextPropertyChanged()
+	static void OnMaskPropertyChanged(BindableObject bindable, object oldValue, object newValue)
 	{
-		if (applyingMask)
-		{
-			return;
-		}
+		var mask = (string?)newValue;
+		var maskedBehavior = (MaskedBehavior)bindable;
 
-		applyingMask = true;
-		ApplyMask(View?.Text);
-		applyingMask = false;
+		maskedBehavior.SetMaskPositions(mask);
 	}
 
-	void SetPositions()
+	static async void OnUnmaskedCharacterPropertyChanged(BindableObject bindable, object oldValue, object newValue)
 	{
-		if (string.IsNullOrEmpty(Mask))
+		var maskedBehavior = (MaskedBehavior)bindable;
+		await maskedBehavior.OnMaskChanged(maskedBehavior.Mask).ConfigureAwait(false);
+	}
+
+	Task OnTextPropertyChanged() => ApplyMask(View?.Text);
+
+	void SetMaskPositions(in string? mask)
+	{
+		if (string.IsNullOrEmpty(mask))
 		{
-			positions = null;
+			maskPositions = null;
 			return;
 		}
 
 		var list = new Dictionary<int, char>();
-		if (Mask != null)
+
+		for (var i = 0; i < mask.Length; i++)
 		{
-			for (var i = 0; i < Mask.Length; i++)
+			if (mask[i] != UnmaskedCharacter)
 			{
-				if (Mask[i] != UnMaskedCharacter)
-				{
-					list.Add(i, Mask[i]);
-				}
+				list.Add(i, mask[i]);
 			}
 		}
 
-		positions = list;
+		maskPositions = list;
 	}
 
-	void OnMaskChanged()
+	async ValueTask OnMaskChanged(string? mask)
 	{
-		if (string.IsNullOrEmpty(Mask))
+		if (string.IsNullOrEmpty(mask))
 		{
-			positions = null;
+			maskPositions = null;
 			return;
 		}
 
-		var originalText = RemoveMaskNullableString(View?.Text);
-		SetPositions();
-		ApplyMask(originalText);
+		var originalText = RemoveMask(View?.Text);
+
+		SetMaskPositions(mask);
+
+		await ApplyMask(originalText);
 	}
 
-	string? RemoveMaskNullableString(string? text)
+	[return: NotNullIfNotNull("text")]
+	string? RemoveMask(string? text)
 	{
-		if (text == null || string.IsNullOrEmpty(text))
+		if (text is null)
 		{
-			return text;
+			return null;
 		}
 
-		return RemoveMask(text);
-	}
+		if (text == string.Empty)
+		{
+			return string.Empty;
+		}
 
-	string RemoveMask(string text)
-	{
-		var maskChars = positions?
+		var maskChars = maskPositions?
 			.Select(c => c.Value)
 			.Distinct()
 			.ToArray();
 
 		return string.Join(string.Empty, text.Split(maskChars));
 	}
 
-	void ApplyMask(string? text)
+	async Task ApplyMask(string? text)
 	{
-		if (text != null && !string.IsNullOrWhiteSpace(text) && positions != null)
-		{
-			if (text.Length > (Mask?.Length ?? 0))
-			{
-				text = text.Remove(text.Length - 1);
-			}
+		await applyMaskSemaphoreSlim.WaitAsync();
```

Comment:
Should we pass cancellation token here?

Reply:
I added the `CancellationToken token = default` as a parameter, but then realized that there's no way to pass in a token from .NET MAUI. I.e. These methods are (currently) only called from `PropertyChanged` which doesn't provide a token. 

But, I still think it's a good idea to have a `CancellationToken` as a parameter so that these methods can be leveraged by future code that maybe can supply a token 👍